docs/design/ha_master.md

```diff
@@ -0,0 +1,188 @@
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all VMs for kubernetes masters (see VM naming section).
+Based on that, we will set the following flags:
+* `-initial-cluster` - list of all hostnames for master machines
+* `-initial-cluster-state` (keep in mind that we are adding masters one by one):
+  * `new` if we are adding the first replica, i.e. the list of existing masters is empty
+  * `existing` if there are more than one replica, i.e. the list of existing masters is non-empty.
+
+This will allow us to have exactly the same logic for HA and non-HA master. List of VMs
+with master replicas will be generated in `kube-up.sh` script and passed to as a env variable
+`INITIAL_ETCD_CLUSTER`.
+
+### apiservers
+
+All apiservers will work independently. Requests to etcd will go to an instance running on
+the machine and will be forwarded by etcd server to the master if necessary (as explained here).
+
+In order to expose a single endpoint the for the user we will create a load balancer in
+front of all apiservers. Load balancer will use the same IP address as the first kubernetes
+apiserver such that all client configuration will not be affected by such change.
+
+Certificates used by apiserver will only include IP address that will be used by load balancer.
+As a result talking directly to apiserver will not work, unless user explicitly specifies
+appropriate flag:
+* `kubectl --insecure-skip-tls-verify=true`
+* `curl --insecure`
+* `wget --no-check-certificate`
+
+Caching mechanism which is implemented in apiserver will not be affected by replicating master because:
+* GET requests go directly to etcd
+* LIST requests go either directly to etcd or to cache populated via watch (depending on the
+ResourceVersion in ListOptions). In the second scenario, after a PUT/POST request, changes
+might not be visible in LIST response. This is however not worse than it is with the current single master.
+* WATCH does not give any guarantees when change will be delivered.
+
+### controller manager, scheduler & cluster autoscaler
+
+Controller manager and scheduler will by default use a lease mechanism to choose an active instance
+among all masters. Only one instance will be performing any operations.
+All other will be waiting in standby mode.
+
+We will use the same configuration in non-replicated mode to simplify deployment scripts.
+
+### add-on manager
+
+All add-on managers will be working independently. Each of them will observe current state of
+add-ons and will try to sync it with files on disk. As a result, due to races, a single add-on
+can be updated multiple times in a row after upgrading the master. Long-term we should fix this
+by using a similar mechanisms as controller manager or scheduler. However, currently add-on
+manager is just a bash script and adding a master election mechanism would not be easy.
```

Comment:
I don't understand what's the issue here. Addons will be updates when add-on (together with master) will be updated. The only problem is that a single addon might be updated and then reverted as we'll have multiple versions of addon manager competing. It will end once all masters are updated.